new macro to check and test for lib and header (#143)

svigerske opened the pull request:
Adds macro
```
AC_COIN_CHK_LIBHDR([prim],[clients],[lflgs],[cflgs],[dflgs],
              [function-body],[includes],
              [dfltaction],[cmdopts])
```
and replaces `AC_COIN_CHK_LIB` by a wrapper around `AC_COIN_CHK_LIBHDR`.

LouHafer committed:
My initial run at the *_LIBHDR macro chain and wrapper version of CHK_LIB.
Various tweaks to coin.m4 to correct some incorrect calls to DEF_PRIM_ARGS.

LouHafer committed:
A little cleanup of failure mode reporting.

svigerske commented on coin_chk_libhdr.m4:
```diff
@@ -0,0 +1,432 @@
+# There's really no reasonable assumptions that can be made about how a data
+# directory is specified, so there's no attempt to test for existence.
+
+# [cmdlineopts] ($8) specifies the set of configure command line options
+# processed: 'nodata' processes --with-prim, --with-prim-lflags,
+# and --with-prim-cflags; 'dataonly' processes only --with-prim and
+# --with-prim-data; anything else ('all' works well) processes all four
+# command line options. Shell code produced by the macro is tailored based
+# on [cmdlineopts]. `nodata' is the default.
```
Would it make sense to split `COIN_FIND_PRIM_LIBHDR` into two macros, one that checks lib and header, and one that checks the datadir? That should allow to get rid of the `cmdlineopts` argument, but I'm not sure how much of `COIN_FIND_PRIM_LIBHDR` would need to be duplicated.

LouHafer replied:
Could be done --- `FIND_PRIM_LIBHDR` would separate fairly cleanly --- but it just pushes the complexity back into `CHK_LIBHDR` unless we create a separate `CHK_DATA` front end.  I guess I don't see a real advantage.

svigerske commented on coin_chk_libhdr.m4:
```diff
@@ -0,0 +1,432 @@
+#       --with-prim-lflags=-lprim
+#       --with-prim-data=/usr/local/share/prim
+#     if the user doesn't override lflags & data from the command line
+#     prim status is set to requested
```
I wonder whether `--with-prim` without argument or `--with-prim=yes` should just give an error.

When the system does some guesses like here, it tries to be helpful, but it can make it difficult to understand what is going on eventually. I find it better if the user has to be explicit.

So the help should suggest `--without-prim`, `--with-prim-lflags`, `--with-prim-cflags`, `-with-prim-data` and we can still interpret `--with-prim=xyz` as `--with-prim-lflags=xyz` for a bit of backward compatibility.

LouHafer replied:
Stumbled over this point while working on `CHK_LAPACK`. I'm inclined to agree. `with-prim` should be a yes/no thing. `with-prim=no` is absolute, any values for `with-prim-lflags` or `with-prim-cflags` are ignored. `with-prim=yes` says use the values of `with-prim-lflags` and `with-prim-cflags`, whatever they might be (including null) with no guessing (but note that the developer can supply default values as macro parameters). If this fails (and it likely will without help from the developer), let it fail. We can then suggest supplying correct values for lflags & cflags.

The way the macros are written, if the developer has set the default for a package to `no`, an explicit `with-prim=yes` is required. Without it, values for `with-prim-lflags` and `with-prim-cflags` are ignored. This might not be the most intuitive approach. I can see an argument that says supplying values for `with-prim-lflags` and `with-prim-cflags` should imply `with-prim=yes`. Can be done, just makes the command line processing logic a bit more complicated.

svigerske replied:
Would be nice if specifying one of the `--with-prim-xyz` automatically enables the package and does not require an extra `--with-prim`.

LouHafer replied:
I was afraid you'd say that  :-)    Done.

svigerske commented on coin_chk_libhdr.m4:
```diff
@@ -0,0 +1,432 @@
+# The usage portion of [dfltaction] ($7) (no, yes) is used as the default
+# value of --with-prim if the user offers no guidance via command line
+# parameters. The (hardwired) default is yes.  See LIBHDR_DFLT_HACK for the
+# full story.
```
I doubt I fully understood the whole story about separate compile and link checks.

But I wonder whether one couldn't infer what to do from whether function-body and/or includes is empty.
- If a function-body is given and is not only a single word, assemble a code that compiles and links from both includes and function-body.
- If function-body is a single word, then do a linkcheck only. If includes are given, raise an error?
- If only includes are given, then check whether that compiles with an empty function-body.

LouHafer replied:
I'm going to argue your philosophical position: don't guess.  Also, the macros are set up in such a way that most users will never need to bother with `[dfltaction]`. The macros assume that if you're asking about a library, the default should be to use it; overriding this requires only `no` as `[dfltaction]`.  Explicitly specifying the link test strategy (`sep` or `tog`) should hardly ever be necessary.

The `CHK_LIB` wrapper macro is defined to handle the common case where the user supplies only a function (i.e., single word) and a header file name and wants separate checks for header usability (check that `#include header` compiles) and ability to link (`void function()` can be linked). It hides the complication of the link test strategy from the user by specifying `sep` but won't override an explicit user spec if it's present in `[dfltaction]` (even though forcing `tog` is almost certainly an error here).

The underlying `CHK_LIBHDR` already does the right thing when either of `[function-body]` or `[includes]` are null. The only ambiguity is when both are non-null. `CHK_LIBHDR` assumes that if both are present, they should be tested as a whole program in the link check. The only time the user needs to intervene is if they're calling `CHK_LIBHDR` directly, both `[function-body]` and `[includes]` are non-null, and they really want separate compile and link tests. (As per above, this is exactly the case that `CHK_LIB` hides from the user.)

svigerske replied:
Could one then at least write out the name of the strategies? `sep` I can still guess, but that `tog` should be short for `together` is not so intuitive. If one reads a configure.ac file, it would be good if one can get some idea what these flags do without having to read up the coin.m4 file and deciphering LIBHDR_DFLT_HACK.

LouHafer replied:
Maybe something like {yes, no} x {separate, together} with a space in between for readability (e.g., `yes separate`). Full words for sure. I'll see if the space works.

LouHafer replied:
Check the new version of `LIBHDR_DFLT_HACK`. It parses strings of space-separated keywords. I made the keywords more human-friendly. As collateral damage, you can now write keywords in any order. Should be easily expansible to handle more overload types and keywords, as long as the keywords are unique.  What we might want to do is change the `yes` and `no` keywords for usage to something like `default_use` and `default_skip`. Might break existing configure.ac, but the `[dfltaction]` parameter of the old `CHK_LIB` seems to be rarely used. No examples that I see in the basic optimisation stack. Any that you know of?

svigerske replied:
That would be a nice change.

I don't remember a place where it is currently used. Usually one wants to use everything there is available.

LouHafer replied:
Done, with the latest commits.

svigerske commented on coin_chk_libhdr.m4:
```diff
@@ -0,0 +1,432 @@
+  m4_ifnblank([$6],
+    [if test $m4_tolower(coin_has_$1) != skipping ; then
+       ac_save_CXXFLAGS=$CXXFLAGS
+       CXXFLAGS="$m4_tolower($1_cflags)"
```
Shouldn't this be `CPPFLAGS`?

The user may have specified `AC_LANG_PUSH(C)` and that C++ flags would be ignored?

svigerske replied:
Yes, CPPFLAGS seems to be the safe choice to me.

LouHafer replied:
Yep. Fixed.

svigerske commented on coin.m4:
```diff
@@ -1870,7 +1870,7 @@ dnl Make sure the necessary variables exist for each client package.
 
 dnl Set up command line arguments with DEF_PRIM_ARGS.
-  AC_COIN_DEF_PRIM_ARGS([lapack],yes,yes,no,no)
+  AC_COIN_DEF_PRIM_ARGS([lapack],yes,yes,yes,no)
```
We don't do anything with `--with-lapack-cflags`.

LouHafer replied:
True, we don't need them for the link check in `CHK_LAPACK` but the user might want to point to a nonstandard place for headers to compile code.

svigerske replied:
But how does the code then know which Lapack implementation it is, i.e., which header file it could include. We don't have a detection or distinction for MKL, OpenBlas, Atlas, etc. so far.

LouHafer replied:
The only place this matters is when the user specifies `--with-lapack-cflags` on the command line. We should respect that. This applies in the case (my test MKL installation, for example) where both libraries and header files are in a nonstandard (non-system) location. All the other tests assume we're looking for something that's properly installed in the usual system locations, hence cflags doesn't really come into play (at least, in terms of `-Ipath/to/include`).

svigerske replied:
But then your test assumes that Lapack is MKL and that you can include mkl.h. If Lapack is something else, even if that something comes with some headers and corresponding cflags are given, that test won't compile.

I had turned off the `--with-lapack-cflags` in 1837ba4a8 to not give the impression in `configure --help` that the lapack cflags would be used for anything.

LouHafer replied:
No argument with what you're saying --- as soon as the user starts using a specific header in their code, they're tied to that implementation of LAPACK. The flip side is that if they do want to choose a specific implementation, specifying `--with-lapack-lflags` and `--with-lapack-cflags` is their only chance. Still, I'm coming around to your point of view. `CHK_LAPACK` does the right thing (link check only) for COIN code, which is careful not to require a specific implementation. If the user is writing their own code and wants a specific implementation, it should be a separate project and they can configure as they like, using `CHK_LIBHDR` & friends. I'll disable `--with-lapack-cflags`.

LouHafer committed:
Changes to COIN_NAMEMANGLING and COIN_TRY_LINK to support 'extra underscore'.

LouHafer committed:
Change TRY_LINK so `checking for' messages are suppressed by default. Change
CHK_LAPACK to look for dorhr_col (embedded underscore, so we get that right)
and produce a readable `checking for Lapack' message.

LouHafer commented:
@svigerske, I've ended up doing a fair bit of tweaking to `CHK_LAPACK`, to better handle configure command line parameters. I can kind of see a general problem coming where lines of code that are commented on here are going to disappear. What's the preferred strategy? Just go ahead an push changes into the issue143 branch?

svigerske commented:
> @svigerske, I've ended up doing a fair bit of tweaking to `CHK_LAPACK`, to better handle configure command line parameters. I can kind of see a general problem coming where lines of code that are commented on here are going to disappear. What's the preferred strategy? Just go ahead an push changes into the issue143 branch?

Yes, that should be no problem.
GitHub should recognize that the comments are for an older version and still show the correct diff (at least that's how it works in GitLab).

LouHafer committed:
Split CHK_LAPACK into its own file for ease of work. Fix it up so that it
does a better job with command line parameters. Adjust CHK_LIB_HDR so that
any of --with-prim, --with-prim-lflags, --with-prim-cflags, --with-prim-data
override a default do not use. Use CPPFLAGS instead of CXXFLAGS for compile
and link checks. Do not guess default values for lflags, data. Add helpful
advice in the event of compile or link failure.

LouHafer commented:
Well piffle! Apologies for polluting the log with all those past commits. I'm still baffled by git some days. Looked much cleaner locally, a nice fast-forward with four commits. Anyway, I'm hoping this addresses all the issues except the question of splitting `PRIM_CHK_HDR`. I'll consider that one after lunch.

LouHafer commented on coin_chk_lapack.m4:
```diff
@@ -0,0 +1,195 @@
+  else
+    lapack_keep_looking=no
+    AC_MSG_RESULT([no (user-specified)])
+  fi
```
The idea here is to pay attention to command line flags, if present. In particular, `--with-lapack=no` will be honored. If the user specifies `--with-lapack-lflags` or `--with-lapack-cflags`, that will be picked up and we'll try the link. `lapack_cflags` isn't actually used in the link test, but it's captured. I added `lapack_keep_looking` just to keep the code structure easy to see (didn't want to introduce a massive arm on an if-then-else statement). Probably should mention that I tweaked `TRY_LINK` to suppress messages by default, so that I could get readable `checking ... result` output here.

svigerske replied:
The TRY_LINK message was saying what it looks for using which linker flags, right? Don't you think that this was useful?

LouHafer replied:
It was definitely useful, just not very readable. I'll grab some comparative output and we can decide. Give me a bit of time for that.

LouHafer replied:
Here's the stdout output on my linux box configuring CoinUtils. With `TRY_LINK` messages turned off.
```
checking for LAPACK... yes (lapack.pc and blas.pc)
```
With `TRY_LINK` output turned on,
```
checking for LAPACK... checking for function dorhr_col in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function dorhr_col_ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function DORHR_COL in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function DORHR_COL_ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function dorhr_col_ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function dorhr_col__ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function DORHR_COL_ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
checking for function DORHR_COL__ in -lmkl_core -lmkl_intel_lp64 -lmkl_sequential -lm ... no
yes (lapack.pc and blas.pc)
checking for function dorhr_col in -llapack  ... no  
checking for function dorhr_col_ in -llapack  ... yes
```
It's informative if you know what you're looking at (the attempt to figure out name mangling, which fails the first 8 times because ld can't find the MKL libraries) but not so much if you're unaware of name mangling and just want to know that some LAPACK was found.

The attempts to figure out name mangling do show in config.log, but (annoyingly) the final successful attempt is not shown, only the failures.

I turned `TRY_LINK` output off by default because it didn't seem to be used anywhere except `CHK_LAPACK`, so default off was minimal change. I can flip that, and explicitly set it to off in `CHK_LAPACK`. Then output will show if TRY_LINK is used elsewhere.

If what you're after is a bit more information, we could print the final results (link flags and name mangling) at the end.

svigerske replied:
Na fine, maybe it was just interesting for me to see.

Just changing `AC_MSG_RESULT([yes (system special case)])` to print `$lapack_lflags`, too, would be nice.

LouHafer replied:
I've tweaked `CHK_LAPACK` to produce a message at the end showing what was found. If `coin_has_lapack = yes` the values of `lapack_lflags` and `lapack_pcfiles` are printed if they're not empty.

LouHafer commented on coin_chk_libhdr.m4:
```diff
@@ -106,19 +106,21 @@ AC_DEFUN([AC_COIN_LIBHDR_DFLT_HACK],
-  m4_tolower($1_data)="m4_default([$4],[/usr/local/share/$1])"
+  m4_tolower($1_lflags)="$2"
+  m4_tolower($1_cflags)="$3"
+  m4_tolower($1_data)="$4"
```
This change removes the guesses for lflags and data.

LouHafer commented on coin_chk_libhdr.m4:
```diff
@@ -139,10 +141,17 @@ dnl to prim_data, otherwise to prim_lflags.
+       test -n "$m4_tolower(with_$1_data)" ; then
+      m4_tolower(coin_has_$1)=requested
+      m4_tolower($1_failmode)=''
+    fi
```
This allows non-null lflags, cflags, or data to override a default do not use.

LouHafer commented on coin_chk_libhdr.m4:
```diff
@@ -354,6 +363,22 @@ dnl to do the heavy lifting.
+	[You can supply correct values using --with-m4_tolower($1)-lflags.]))
+    fi
+  fi
+
```
Helpful advice. A bit fragile --- if the failure modes set in FIND_PRIM_LIBHDR change, this can break.

svigerske replied:
Users often don't recall that they should look into config.log regarding details why a compile or link check failed. Maybe a hint towards config.log could be added here, too.

LouHafer replied:
Will do.

LouHafer reviewed (commented):
Hmmm ... maybe I should have clicked 'leave a single comment' instead of 'start a review.'

LouHafer committed:
Remove all mention of compiler flags from CHK_LAPACK. CHK_LAPACK is
specifically designed to confirm only the ability to link.

LouHafer committed:
Add a hint to the user to check config.log for details of compile or link
failures.

LouHafer committed:
Rewrite LIBHDR_DFLT_HACK to handle strings of space-separated keywords. Make
the keywords more human-friendly.

LouHafer committed:
Push all result messages to the end and print something informative.

LouHafer committed:
Minor bit of cleanup in TRY_LINK. m4_ifnblank is more restrictive than test -n.

tkralphs commented:
Before we merge this, let's clean up the history a bit. Lou, I guess you are pulling master with the default merging behavior. You need to pull with rebase in order to avoid all of those old commits coming in. You will need to do a force push to your fork after rebasing, but this shouldn't cause any issues for branches where you are the only contributor. 

(You can make pulling with rebase your default with `git config --global pull.rebase true`, which Is a good idea in many cases)

LouHafer committed:
Changes to COIN_DEF_PRIM_ARGS, LIBHDR_DFLT_HACK, FIND_PRIM_LIBHDR,
CHK_LIBHDR, and CHK_LIB to change the usage keywords from {yes, no, build}
to {default_use, default_skip, default_build}. Also added an m4_fatal trap
to catch attempts to use default_build in the CHK_LIB / CHK_LIBHDR chain
at run_autotools (so the developer can change it). Only CHK_PKG recognises
this option.

LouHafer committed:
Properly handle the cases cmdopt = all and default.

LouHafer committed:
Pulled CHK_PKG and FIND_PRIM_PKG from coin.m4 into their own file. Did a
complete rewrite of both to match the behaviour of the LIBHDR chain. Fixed
dfltaction = build as best it can be fixed. Added warnings that appear
from run_autotools to prompt maintainers to quit using it.

LouHafer committed:
Added a macro to handle the warning / fatal error for using
dfltaction=build. From the LIBHDR chain, this triggers a fatal error at
run_autotools execution. From the PKG chain, it's just a warning.

LouHafer commented:
@svigerske, here's the mass of changes I mentioned in email. I've pulled `COIN_CHK_PKG` and `COIN_FIND_PRIM_PKG` into a separate file for ease of working and rewritten them to conform to the style and algorithms of the LIBHDR chain. `dfltaction = build` will work for simple cases in the PKG chain, and there's now a big ugly macro, `COIN_DEPRECATE_BUILD_OPTION`, in `coin_chk_libhdr.m4` that will force a fatal error if the LIBHDR chain sees 'build' and issue a warning if the PKG chain sees 'build'. This happens at run_autotools time --- only the developer can do anything to fix the problem. There are a bunch of other smaller changes. Have a look and try them on your favourite project and let me know what breaks, or what you don't much care for.

LouHafer commented:
@svigerske, doing a drive-by check and noticed the `thumbs up'. Let me know when (if) you think we're ready to try and merge this and I'll do up a clean pull request. Given the changes, we may well see the occasional bit of breakage in existing configure.ac and Makefile.am. If there are any projects you think might be at risk, let me know. I should have the odd day over the next few weeks and can try to test.

svigerske commented:
I think the changes are ready to be tried. Not sure I am ready, too. We shouldn't merge before trying them. We can do this in separate branches of your and my favorite projects to see what Teds travis and appveyor tests are saying.

svigerske committed:
update comment regarding build-option being deprecated

svigerske committed:
tabs to spaces for correct alignment

svigerske committed:
revise lapack check

- echo flags used within check-result line to avoid strange configure output
- echo user-lapack flags also if error
- remove sgi handling for lapack (no way to test this)

svigerske committed:
bring back taking lapack lflags from --with-lapack

svigerske committed:
bring back messages from AC_COIN_TRY_LINK by default

- suppress in lapack check only

svigerske committed:
avoid pkg-config error-message from empty argument

- e.g., argument 3 of
  AC_COIN_TRY_LINK([pardiso_ipopt_newinterface],[$pardiso_lflags $lapack_lflags],[$lapack_pcfiles],...)
  is not empty, but $lapack_pcfiles may be

svigerske committed:
back to testing for dsyev instead of dorhr_col

- dorhr_col is Lapack 3.9 (released 11/21/19,
  https://github.com/Reference-LAPACK/lapack-www/blob/master/lapack-3.9.0.txt),
  which hasn't been spread  around much yet and COIN-OR projects typically
  don't require such a new version

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/coin-or-tools/BuildTools?pullRequest=148) <br/>All committers have signed the CLA.

svigerske commented:
I tried AC_COIN_CHK_LIBHDR in this form:
```
AC_LANG_PUSH(C++)
AC_COIN_CHK_LIBHDR(SoPlex,[OsiSpxLib OsiTest],[],[],[],
  [new soplex::DIdxSet()],
  [#include "soplex.h"
   #if SOPLEX_VERSION >= 400
   #error "OsiSpx requires SoPlex < 4.0"
   #endif
  ],
  [default_skip])
AC_LANG_POP(C++)
```
and it works when I specify no flags for SoPlex, flags for SoPlex 3.1.1, and flags for current SoPlex for configure.
In the latter case, "works" means that I get
```
checking for library SoPlex with combined link and compile check... no (header compile, link with header)
configure: WARNING:
Compiler flags are "-I/path/to/soplex/src". If you expected this test to
succeed, check that they are correct. You can supply correct values using
--with-soplex-cflags.
configure: WARNING:
Linker flags are "-L/path/to/soplex/lib -lsoplex". If you expected this test to
succeed, check that they are correct. You can supply correct values using
--with-soplex-lflags.
configure: WARNING:
Check config.log for details of failed compile or link attempts.
```
(I find all these linebreaks too much, but ok).

My concern is that configure keeps running, finishing with
```
configure: Configuration of Osi successful
```
I think it should stop with an error if the flags that the user provided explicitly do not work.

svigerske commented:
On Cbc, there is
```
AC_COIN_CHK_PKG([Nauty],[CbcLib CbcSolverLib],[nauty])
```
I configure with `with_nauty_lflags=-lnauty`, which are working flags.

But the macro seems to record `nauty` as a .pc file that should be used for CbcLib, etc, i.e.,
```
CBCLIB_PCFILES='osi-cplex nauty osi-clp clp cgl osi coinutils '
```
and this gives me a number of warnings from the finalize_flags part in configure:
```
Package nauty was not found in the pkg-config search path.
Perhaps you should add the directory containing `nauty.pc'
to the PKG_CONFIG_PATH environment variable
Package 'nauty', required by 'virtual:world', not found
Package nauty was not found in the pkg-config search path.
Perhaps you should add the directory containing `nauty.pc'
to the PKG_CONFIG_PATH environment variable
Package 'nauty', required by 'virtual:world', not found
Package nauty was not found in the pkg-config search path.
Perhaps you should add the directory containing `nauty.pc'
to the PKG_CONFIG_PATH environment variable
Package 'nauty', required by 'virtual:world', not found
Package nauty was not found in the pkg-config search path.
Perhaps you should add the directory containing `nauty.pc'
to the PKG_CONFIG_PATH environment variable
Package 'nauty', required by 'virtual:world', not found
```
(and linking fails)

If a user specifies some `--with-xyz-*` flag, then none of the defaults passed to `AC_COIN_CHK_PKG` should be used, I believe.

svigerske commented:
> @svigerske, doing a drive-by check and noticed the `thumbs up'. Let me know when (if) you think we're ready to try and merge this and I'll do up a clean pull request. Given the changes, we may well see the occasional bit of breakage in existing configure.ac and Makefile.am. If there are any projects you think might be at risk, let me know. I should have the odd day over the next few weeks and can try to test.

I've now tried this on Ipopt, CoinUtils, Osi, and Cbc in branches configure-update. In the configure.ac, I removed the use of `[build]` and replaced `AC_COIN_CHK_LIB` by `AC_COIN_CHK_LIBHDR`.   
That works on my machine now (if I turn off nauty, see above). travis/appveyor builds can be found at
- https://travis-ci.com/github/coin-or/Ipopt
- https://ci.appveyor.com/project/coin-or/ipopt
- https://travis-ci.com/github/coin-or/CoinUtils
- https://ci.appveyor.com/project/coin-or/coinutils
- https://travis-ci.com/github/coin-or/Osi
- https://ci.appveyor.com/project/coin-or/osi
- https://travis-ci.com/github/coin-or/Cbc
- https://ci.appveyor.com/project/coin-or/cbc
(make sure to look at the build for the `configure-update` branch)

We also need to update the docu in [docs/configure.md](https://github.com/coin-or-tools/BuildTools/blob/issue143/docs/configure.md) before merging.

svigerske committed:
remove AC_COIN_NAMEMANGLING

- will no longer be used and can be replaced by a AC_COIN_TRY_LINK call,
  e.g., as in coin-or/Ipopt@eb7c15a43842dedae8c472f3a1e915173e8258ef

svigerske committed:
first check with underscore, then do without underscore

- since having an underscore seems to be the default on my system for
  fortran codes (gfortan 10)

LouHafer committed:
Add a variable prim_userflags in PRIM_FIND_LIBHDR and CHK_LIBHDR to track
the presence of user flags and force a stop on error if they don't work.

LouHafer committed:
In FIND_PRIM_PKG, clear prim_pcfiles when the user supplies flags on the
command line.

LouHafer commented:
@svigerske, I've made a pair of small changes as suggested in your comments. CHK_LIBHDR issues the same warning message as before, then forces an error stop if the failure results from user-supplied flags. FIND_PRIM_PKG now clears prim_pcfiles if the user supplies any flags. See if this does it for you.

LouHafer commented:
@svigerske, I hate to bring up a new glitch, but as I got to rewriting configure.md I ended up looking at COIN_CHK_HERE, and now I'm looking at commit https://github.com/coin-or-tools/BuildTools/commit/5fae46779a7349c285646e1267579470dfae828c. Seems to me the check
```
if test -n "$m4_toupper(myvar)_PCFILES" 
```
is the wrong check, it should consider `m4_default($3, m4_tolower($1))`, which is the name of the .pc file being added. More to the point, this can never be null, so  there's no reason to check at all.

Which got me to thinking, 'What if I want to use CHK_HERE for a primary that doesn't get installed? No reason that it should have a .pc file, it's not useful.' So maybe CHK_HERE really does need to consider the possibility that there is no .pc file. The change would not be backward compatible. There are surely places where the default is used. Osi2, for one. Still, if we're going to change, this is as good a time as any.

LouHafer committed:
Fix the test of COIN_SKIP_PROJECTS so it actually works.

LouHafer commented:
So fixing up the documentation is a good thing. https://github.com/coin-or-tools/BuildTools/pull/148/commits/1f157211e0b91e09594d134e5fef35824e3d9747 because I thought maybe it'd be good to actually check that COIN_SKIP_PROJECTS works  :-)

LouHafer committed:
First cut at a complete rewrite of configure.md for the new versions of
CHK_PKG and CHK_LIBHDR.

LouHafer commented:
@svigerske, I've done a major rewrite of configure.md. See what you think.

svigerske committed:
tabs to spaces again

svigerske committed:
change error message

- before:
  configure: error: Aborting configure; user-specified flags failed.
- new:
  configure: error: user-specified flags for SoPlex do not work.
- since this is the last message, I think it is pretty clear that
  configure aborted

svigerske committed:
remove wrong test

- meant to test whether m4_default($3,m4_tolower($1)) is not empty
- and it cannot be empty

svigerske commented:
> @svigerske, I hate to bring up a new glitch, but as I got to rewriting configure.md I ended up looking at COIN_CHK_HERE, and now I'm looking at commit [5fae467](https://github.com/coin-or-tools/BuildTools/commit/5fae46779a7349c285646e1267579470dfae828c). Seems to me the check
> 
> ```
> if test -n "$m4_toupper(myvar)_PCFILES" 
> ```
> 
> is the wrong check, it should consider `m4_default($3, m4_tolower($1))`, which is the name of the .pc file being added. More to the point, this can never be null, so there's no reason to check at all.

Yes, right. I fixed that in this branch now (67267ca).
AC_COIN_CHK_HERE was not much used and the other two places from 5fae467 were ok.

> Which got me to thinking, 'What if I want to use CHK_HERE for a primary that doesn't get installed? No reason that it should have a .pc file, it's not useful.' So maybe CHK_HERE really does need to consider the possibility that there is no .pc file. The change would not be backward compatible. There are surely places where the default is used. Osi2, for one. Still, if we're going to change, this is as good a time as any.

Yes, would be ok to change this now.

svigerske committed:
tests don't use RPATH_FLAGS

- because they typically rely on libtool, too

svigerske committed:
make clear that COIN_HAS_SUB are automake-conditionals and not variables

- the variables would be COIN_HAS_SUB_TRUE and COIN_HAS_SUB_FALSE

svigerske committed:
remove that comment about not having to check for -lm if using coinutils

- projects may still require an explicit -lm on the linkline, which they
  don't get just from the coinutils so depending on libm

svigerske commented:
> @svigerske, I've done a major rewrite of configure.md. See what you think.

Looks very good. Thank you!

svigerske commented:
So, to me, this looks good now. My machine and travis+appveyor were ok on the corresponding branches of Ipopt, CoinUtils, Osi, and Cbc.

@LouHafer Can we just merge this, or you want to work a bit more on `AC_COIN_CHK_HERE` first?

LouHafer commented:
@svigerske, it's true that the issue with `COIN_CHK_HERE` is a solution in search of a problem. It can be dealt with later if anyone complains. Is it easy for you to do a clean pull from this branch in spite of the excess history I dragged in? If not, say so and I'll cobble together a clean pull request.

svigerske commented:
> @svigerske, it's true that the issue with `COIN_CHK_HERE` is a solution in search of a problem. It can be dealt with later if anyone complains. Is it easy for you to do a clean pull from this branch in spite of the excess history I dragged in? If not, say so and I'll cobble together a clean pull request.

So I did a `git rebase -i master` and removed commits from the past that had already been cherry-picked into master or were shown twice (probably because of this merging back-and-forth). That led me rebase with almost no conflicts and the final result looks the same (empty diff).
